pkg/build/strategies/sti/sti.go

```diff
@@ -591,16 +590,17 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
 	go dockerpkg.StreamContainerIO(errReader, &errOutput, func(a ...interface{}) { glog.Info(a...) })
 
 	err := builder.docker.RunContainer(opts)
-	if util.IsTimeoutError(err) {
-		// Cancel waiting for source input if the container timeouts
-		wg.Done()
-	}
 	if e, ok := err.(errors.ContainerError); ok {
 		// even with deferred close above, close errReader now so we avoid data race condition on errOutput;
 		// closing will cause StreamContainerIO to exit, thus releasing the writer in the equation
 		errReader.Close()
 		return errors.NewContainerError(config.BuilderImage, e.ErrorCode, errOutput)
 	}
+	// Do not wait for source input if the container times out.
+	// FIXME: this potentially leaks a goroutine.
+	if !util.IsTimeoutError(err) {
+		wg.Wait()
+	}
```

Comment:
@bparees thanks for challenging my tired mind :)
Really, good to have another pair of eyes and good brain to reason about it.

> 1) injections start (wg.Add(1) occurs)
> 2) call to start container times out (container never starts)

This means that after

```go
err := builder.docker.RunContainer(opts)
```

`util.IsTimeoutError(err) == true`, right?


> 3) injections never finish because the container never started

Yes, and this causes a goroutine leak.

> 4) wg.Done() never gets called because the injections never finish
> 5) wg.Wait() never completes

When `util.IsTimeoutError(err) == true`, `wg.Wait()` is not called. In all other cases, i.e. `!util.IsTimeoutError(err)`, `wg.Wait()` is called.